test/external_project/CMakeLists.txt

```diff
@@ -0,0 +1,90 @@
+endif ()
+
+# 1) find_package: seqan3 is a submodule and we add it with add_subdirectory
+#    This should work with cmake >= 3.4.
+# (ExternalProject_Add simulates a fresh and separate invocation of cmake ../)
+ExternalProject_Add (
+    seqan3_submodule_add_subdirectory
+    PREFIX seqan3_submodule_add_subdirectory
+    SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/seqan3_submodule_add_subdirectory"
+    CMAKE_COMMAND "${SEQAN3_EXTERNAL_PROJECT_CMAKE_COMMAND}"
+    CMAKE_ARGS
+        ${SEQAN3_EXTERNAL_PROJECT_CMAKE_ARGS}
+        "-DSEQAN3_ROOT=${SEQAN3_ROOT}"
+        "-DCMAKE_FIND_DEBUG_MODE=${SEQAN3_EXTERNAL_PROJECT_FIND_DEBUG_MODE}"
+)
+
+# 2) find_package: seqan3 is a submodule and we only specify the path to `seqan3-config.cmake`
+#    This should work with cmake >= 3.4.
+# (ExternalProject_Add simulates a fresh and separate invocation of cmake ../)
+ExternalProject_Add (
+    seqan3_submodule_find_package
+    PREFIX seqan3_submodule_find_package
+    SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/seqan3_submodule_find_package"
+    CMAKE_COMMAND "${SEQAN3_EXTERNAL_PROJECT_CMAKE_COMMAND}"
+    CMAKE_ARGS
+        ${SEQAN3_EXTERNAL_PROJECT_CMAKE_ARGS}
+        "-DCMAKE_PREFIX_PATH=${SEQAN3_ROOT}/build_system"
+        "-DCMAKE_FIND_DEBUG_MODE=${SEQAN3_EXTERNAL_PROJECT_FIND_DEBUG_MODE}"
+)
+
+# 3) find_package: seqan3 is system-wide installed / PREFIX_INSTALL (unzipped)
+#    This should work with cmake >= 3.4, but we need cmake >= 3.14 to be able to install seqan3.
+# (ExternalProject_Add simulates a fresh and separate invocation of cmake ../)
+if (NOT ${CMAKE_VERSION} VERSION_LESS 3.14) # cmake 3.14 version is needed to install seqan3.
+    ExternalProject_Add (
+        seqan3_installed
+        PREFIX seqan3_installed
+        SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/seqan3_installed"
+        CMAKE_COMMAND "${SEQAN3_EXTERNAL_PROJECT_CMAKE_COMMAND}"
+        CMAKE_ARGS
+            ${SEQAN3_EXTERNAL_PROJECT_CMAKE_ARGS}
+            "-DCMAKE_SYSTEM_PREFIX_PATH=${SEQAN3_SYSTEM_PREFIX}"
+            "-DCMAKE_FIND_DEBUG_MODE=${SEQAN3_EXTERNAL_PROJECT_FIND_DEBUG_MODE}"
+    )
+    add_dependencies(seqan3_installed seqan3_test_prerequisite)
+endif ()
+
+# 4) find_package: seqan3 was downloaded as an archive file via fetch content (e.g. zip, tar.xz)
+#    FetchContent is a cmake >= 3.14 feature.
+# (ExternalProject_Add simulates a fresh and separate invocation of cmake ../)
+if (NOT ${CMAKE_VERSION} VERSION_LESS 3.14) # cmake 3.14 version is needed to use fetch_content.
+    ExternalProject_Add (
+        seqan3_fetch_content_zip
+        PREFIX seqan3_fetch_content_zip
+        SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/seqan3_fetch_content_zip"
+        CMAKE_ARGS
+            ${SEQAN3_EXTERNAL_PROJECT_CMAKE_ARGS}
+            "-DSEQAN3_PACKAGE_ZIP_URL=${SEQAN3_PACKAGE_ZIP_URL}"
+            "-DCMAKE_FIND_DEBUG_MODE=${SEQAN3_EXTERNAL_PROJECT_FIND_DEBUG_MODE}"
+    )
+    add_dependencies(seqan3_fetch_content_zip seqan3_test_prerequisite)
+endif ()
```

Comment:
From looking at the CI,

`seqan3_installed` is run with cmake 3.4.3 because of the `CMAKE_COMMAND`
What is the command about 3.14 needed then?

`seqan3_fetch_content_zip` is run with cmake 3.14 because there is no `CMAKE_COMMAND`


Maybe you can write a few sentences about what is tested with what?
Why do we not test `seqan3_submodule_add_subdirectory` and `seqan3_submodule_find_package` also with cmake 3.14? is it guaranteed to not break?

Reply:
> `seqan3_installed` is run with cmake 3.4.3 because of the `CMAKE_COMMAND`

Yes



> What is the comment about 3.14 needed then?

Our install script of seqan3 needs cmake version 3.14.



> `seqan3_fetch_content_zip` is run with cmake 3.14 because there is no `CMAKE_COMMAND`

Correct

> Maybe you can write a few sentences about what is tested with what?

Okay, I'm not completely sure how to describe it, but I'll try



> Why do we not test `seqan3_submodule_add_subdirectory` and `seqan3_submodule_find_package` also with cmake 3.14? is it guaranteed to not break?

We could, I just want to make sure that our library is working with 3.4 as minimum requirement. Cmake is pretty good with compatibilities between version. I never encountered breakage between versions.